Fixes #14395 - Label for provision template capitalized in os details…

xprazak2 opened the pull request:
… page

Commit:
Fixes #14395 - Label for provision template capitalized in os details page

s

domcleal commented:
This is going to make the UI inconsistent, especially "fixing" it on only one page.  Do you intend to change all pages, and all template kinds?  What about template kinds added in plugins?

You're also opting to translate this one string, but not the others (also [ticket #7171](http://projects.theforeman.org/issues/7171))...

xprazak2 commented:
We now have a pretty name for each template kind. It can be whatever we want and it gets translated. Same goes for templates from plugins. I've made changes in hosts/show, provisioning_templates/index, provisioning_templates/edit, operatingsystems/edit pages to display pretty names, did I miss any?

tbrisker commented on app/helpers/hosts_helper.rb:
```diff
@@ -270,7 +270,7 @@ def show_templates
       templates.map do |kind, tmplt|
         content_tag :tr do
-          content_tag(:td, _("%s Template") % kind) +
+          content_tag(:td, kind + " " + _("Template")) +
```
This will not be translated correctly. In some languages the kind would need to appear after the `Template`. Also `_(kind)` should be used so the kind is translated properly (see below)

tbrisker commented on app/services/template_kind_name_jar.rb:
```diff
@@ -0,0 +1,21 @@
+               "script" => _("Script"),
+               "user_data" => _("User Data"),
+               "ZTP" => _("ZTP"),
+               "POAP" => _("POAP")
```
These will not get correctly translated, use `N_("...")` instead as there is no locale when creating this hash and the translations are evaluated..

xprazak2 commented:
I changed `_("...")` to `N_("...")` and the name on hosts/show page should be translated together with with 'template'.

tbrisker commented on app/helpers/hosts_helper.rb:
```diff
@@ -270,7 +270,7 @@ def show_templates
       templates.map do |kind, tmplt|
         content_tag :tr do
-          content_tag(:td, _("%s Template") % kind) +
+          content_tag(:td,  _("%s Template") % kind) +
```
you added a space but did not extract the kind for translation. Why not move the template part also to the pretty name?

tbrisker commented on app/models/template_kind.rb:
```diff
@@ -6,4 +6,8 @@ class TemplateKind < ActiveRecord::Base
   validates :name, :presence => true, :uniqueness => true
   scoped_search :on => :name
+
+  def pretty_name
```
I think to_label would be a better name for this

tbrisker replied:
Or even possibly just override `.to_s`

tbrisker commented on app/services/foreman/plugin.rb:
```diff
@@ -309,5 +309,10 @@ def provision_method(name, friendly_name)
       Facets.register(klass, name, &block)
     end
+
+    # add human readable label for plugin's template kind with i18n support: template_labels "kind_name" => _("Nice Name")
```
missed the `N_(` here

tbrisker commented on app/helpers/hosts_helper.rb:
```diff
@@ -254,7 +254,7 @@ def show_templates
         template = @host.provisioning_template(:kind => k.name)
         next if template.nil?
-        [k.name, template]
+        [k.pretty_name, template]
```
This will cause trouble for the review link below

tbrisker commented:
@xprazak2 `show_templates` looks like it could use quite a bit of refactoring. I was thinking about moving the first part (fetching the templates) to the controller, and making the second part (all the table) a partial. Also, I'm not sure why we need the whole hash mapping types to templates, each template already has a `template_kind` we can access (and also via that get to the `pretty_name`/`to_label` method).

tbrisker commented on app/services/template_kind_name_jar.rb:
```diff
@@ -0,0 +1,21 @@
+class TemplateKindNameJar
```
Why make this a separate class and not part of `TemplateKind` model?

tbrisker commented on app/services/template_kind_name_jar.rb:
```diff
@@ -0,0 +1,21 @@
+  end
+
+  def self.add(hash)
+    hash.each do |key, value|
```
you can use `Hash#merge!` with a block to achieve the same result without looping over the keys. see http://ruby-doc.org/core-2.2.0/Hash.html#method-i-merge-21 fore more info.

xprazak2 commented:
I moved the jar into the `TemplateKind`, renamed `pretty_name` to `to_label`, refactored `show_templates` - the table is now a partial, templates are supplied from the controller.

tbrisker commented on app/models/template_kind.rb:
```diff
@@ -6,4 +6,25 @@ class TemplateKind < ActiveRecord::Base
+    jar.merge!(hash) { |key| raise Foreman::Exception.new(N_("Cannot add template with key %s, it already exists"), key) }
+  end
+
+  def to_label
```
If you change this to `to_s` you can get rid of all the `.try(:to_label)` later since `nil.to_s => ""` and `to_s` is called implicitly when trying to print the object.

tbrisker commented on app/views/hosts/_templates.html.erb:
```diff
@@ -0,0 +1,25 @@
+    <tbody>
+      <% @templates.each do |tmpl| %>
+        <tr>
+          <td><%= (_("%s Template") % tmpl.template_kind.to_label) %></td>
```
Why not move the `" Template"` part into the jar as well? that way you can make sure the translators get the correct string to translate, e.g. "Provision template" instead of translating "Provision" and "Template" separately, which may have a different translation.

tbrisker commented on app/views/hosts/_templates.html.erb:
```diff
@@ -0,0 +1,25 @@
+          <td><%= (_("%s Template") % tmpl.template_kind.to_label) %></td>
+          <td><%= action_buttons(display_link_if_authorized(_("Edit"), hash_for_edit_provisioning_template_path(:id => tmpl.to_param),
+                                   :rel => "external"),
+                                link_to(_("Review"), url_for(:controller => '/unattended', :action => 'host_template',
```
Nitpick: indentation is a bit off here

tbrisker commented on app/controllers/hosts_controller.rb:
```diff
@@ -272,7 +272,8 @@ def resources
 
   def templates
-    render :text => view_context.show_templates
+    @templates = find_templates
```
You could do this as a `before_filter`

tbrisker commented on app/helpers/hosts_helper.rb:
```diff
@@ -245,44 +245,6 @@ def name_field(host)
-      end
-    end
-
-    return _("No template found") if templates.empty?
```
I think you are missing this line in the refactoring - no templates will lead to an empty table instead of a text explanation.

tbrisker replied:
this as well.

tbrisker commented on app/helpers/hosts_helper.rb:
```diff
@@ -245,44 +245,6 @@ def name_field(host)
   end
 
-  def show_templates
-    unless SETTINGS[:unattended] and @host.managed?
```
You are missing this option in the refactoring.

tbrisker replied:
This is still missing.

tbrisker replied:
nevermind, I saw what you did here

tbrisker commented on app/views/hosts/_templates.html.erb:
```diff
@@ -0,0 +1,25 @@
+<% if @templates.is_a? Array %>
+  <table class="table table-bordered table-striped">
```
Please use `table_css_classes` helper

tbrisker commented:
@xprazak2 thanks, this is looking much better already. Added a few comments inline.

xprazak2 commented:
Main changes: I added the missing options mentioned above, renamed `to_label` ~> `to_s`, added `" Template"` into the jar. Note that it has effect also on operatingsystems/edit and templates/index pages:

![](http://i.imgur.com/MrSNPoE.png)
![](http://i.imgur.com/TNEgB00.png)

lzap commented:
I see you sorted the templates in the controller, but they are not ordered on the screenshot. There has been already a request to sort them and I have attempted to do this, but it was not possible without larger refactoring (https://github.com/theforeman/foreman/pull/1801) - could you make it happen in this patch?

tbrisker commented:
@xprazak2 please rebase, and see if addressing @lzap 's comment is possible to easily achieve here or not.

xprazak2 commented:
I found a way to sort the templates by kind alphabetically, but is this really what we want?

It changes the current order and I kind of fail to see the benefit. If we used pagination, alphabetical ordering would make total sense, but there is a relatively small number of kinds (~10, atm). They can all be comfortably viewed in one tab with the new arrivals displayed at the bottom as it is now. Besides, it will be sorted only for English. I can do sorting in all languages (not sure about results of sorting on Chinese or Korean) but that could mean different order of dropdowns for each language based on translation, which just seems wrong to me.

![](http://i.imgur.com/5AjUdSg.png)

tbrisker commented on app/views/hosts/_templates.html.erb:
```diff
@@ -0,0 +1,38 @@
+<% if @host.managed? %>
```
This should be in the controller, before attempting to find all the templates for an unmanaged host.
You are also missing the `SETTINGS[:unattended]` part.

xprazak2 replied:
`SETTINGS[:unattended]` is unnecessary here, see my [comment](https://github.com/theforeman/foreman/pull/3376#discussion_r60204301) below

tbrisker commented on app/views/hosts/_templates.html.erb:
```diff
@@ -0,0 +1,38 @@
+<% if @host.managed? %>
+  <% if @templates.is_a? Array %>
+    <% if @templates.empty? %>
+      <div><%= _("No template found") %></div>
```
why not pass this from the controller like you do for exceptions?

tbrisker commented on app/views/hosts/show.html.erb:
```diff
@@ -46,9 +46,11 @@
-      <div class="tab-pane" id="template" data-ajax-url='<%= templates_host_path(@host)%>'>
-        <%= spinner(_('Loading template information ...')) %>
-      </div>
+      <% if SETTINGS[:unattended] %>
```
This means that the behaviour will change for `SETTINGS[:unattended]` - instead of a message this div will not be displayed at all.

xprazak2 replied:
No, the behaviour does not change, look [here](https://github.com/theforeman/foreman/blob/develop/app/views/hosts/show.html.erb#L26). I just chose not to render content that cannot be seen. Or do we want a change to display 'templates' tab with a message when `unattended: false` ?

tbrisker replied:
would it make sense to also show this tab only for managed hosts?

tbrisker commented:
@xprazak2 I am also not sure this template sorting is better, in any case since it's not a trivial change and not directly related to this I would say split it off to a different PR. 
Aside from that just a couple of minor things inline and I think we will be good to go.

xprazak2 commented:
I reverted the sorting, we can do it in separate PR if we really want it. I moved most of the checks into controller and handle things like 'no templates found' there with `proxess_ajax_error`, which adds red box around the text (before we had only simple text). When `unattended: false`, templates tab is not displayed at all (same as before).

tbrisker commented:
@xprazak2 Looks good other then my one comment about considering hiding the template tab for unmanaged hosts as well, let me know what you think.

xprazak2 commented:
Sure, seems reasonable. We can hide the templates tab for unmanaged hosts as well. I will make a change.

xprazak2 commented:
'templates tab' is now visible only if `SETTINGS[:unattended] && @host.managed?`

tbrisker commented on app/controllers/hosts_controller.rb:
```diff
@@ -891,4 +892,16 @@ def update_multiple_proxy(proxy_type, host_update_method)
+
+  def find_templates
+    find_resource
+    if @host.managed?
```
This will always be `true` since you added this condition to the parent view

tbrisker commented:
@xprazak2 you forgot to remove the same condition from the controller, once you do we are ready to go.

xprazak2 commented:
Right, sorry about that. Should be ok now.

tbrisker commented:
Looks like we got a nice refactoring out of this seemingly trivial bug. :+1:

tbrisker commented:
Merged as 1ef8f52b525294df333cd628896f8fe0243ad0a8, thanks @xprazak2!

domcleal commented:
@xprazak2 could you please document the new plugin API at http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin?

xprazak2 commented:
There is a new section on our wiki, Translating Template Kind.
